src/Storages/ExternalStream/StorageExternalStream.cpp

```diff
@@ -25,18 +25,42 @@ namespace ErrorCodes
 {
 extern const int NOT_IMPLEMENTED;
 extern const int BAD_ARGUMENTS;
+extern const int INCORRECT_NUMBER_OF_COLUMNS;
+extern const int TYPE_MISMATCH;
 }
 
 namespace
 {
+ExpressionActionsPtr
+buildShardingKeyExpression(const ASTPtr & sharding_key, ContextPtr context, const NamesAndTypesList & columns)
```

Comment:
`ASTPtr sharding_expr`

Reply:
I used `ASTPtr` at the first place. But I found that there are a lot places use `const ASTPtr &`, in fact, `StorageStream.cpp` uses `const ASTPtr &`, thus I changed it. When it should use `const ASTPtr &`? In this same file, line 132 below, in the `StorageExternalStream` ctor, I also used `const ASTPtr &`, should I change it too?

Reply:
This is a v good question. First of all, always stick to `const ASTPtr &` whenever possible since it doesn't need to bump up the atomic ref count. In this case, we shall stick to `const ASTPtr & ` if the `TreeRewrtier` / `ExpressionAnalzyer` accept a const ref. If they don't, for this `buildShardingKeyExpression(...)` specifically, it is good to just accept `ASTPtr` like we did here after change, it simplify caller and as well as the callee.

Reply:
This is a general advice / guideline for `ContextPtr` etc as well.